### DISCUSSION TITLE
The Capacity Issue of Usedbydataset and Refquota Attributes in ZFS FileSystem dataset.

### DISCUSSION THREAD

**Discussion by hhb777** [openzfs/zfs#16727] [Q&A] [answered]
Hello, I have a question to ask.
zfs version: 2.1.6
A 200GB file system dataset was created, and after the data was fully written, it was found that the usedbydataset attribute value of the file system was larger than the size set by refquota, why? There are no snapshots or sub datasets available.
As shown in the screenshot below:
![image](https://github.com/user-attachments/assets/30ce325a-0ea1-403e-931c-811d52624b3f)
The attributes are as follows:
![image](https://github.com/user-attachments/assets/88cb83f2-83be-4522-96ba-eed2cca6c56a)
![image](https://github.com/user-attachments/assets/45783f81-47a3-48e9-87f9-3d8f0e4d0fae)
![image](https://github.com/user-attachments/assets/9c49e653-d563-4c5b-8964-b36abb2b578b)

Thank You!

**Comment by IvanVolosyuk** [comment]
I can confirm that. I did an experiment:
zfs create ...
zfs set refquota=1G ...
dd if=/dev/urandom bs=100M of=...

...and got the difference of 40k for 128k recordsize and 28M for 4k
recordsize (created file was 8M larger).

I would imagine that it is hard to do that 100% correctly going through
many abstraction layers in ZFS, thus it is possible to slightly overshoot.

**Reply by amotin** [reply, marked as answer]
IIRC some overshoot possibility was added there intentionally not too long ago.  The problem is that ZFS does not know how much space each specific operation will take until it completes pool-wide transaction group commit.  Previously reaching quota triggered extra commit.  It resulted in more precise quota behavior, but badly hurt pool performance for other datasets/users.  Instead now ZFS allows to go slightly (up to 3%) over the quota.  See: https://github.com/openzfs/zfs/pull/13839 .

**Comment by hhb777** [comment]
Thank you for your reply. Reference # 13839 can solve my problem. Thank you!
